Update deprecated mem_ops.h function calls to use std::span-based versions and resolve some TODOs

KaganCanSit opened the pull request:
Hello,

This draft PR includes updates to modernize the codebase by replacing deprecated mem_ops.h function calls with their `std::span`-based counterparts in line with C++20 standards. Additionally, I also discussed some TODO comments that I thought would be quick fixes.

My reference points the [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) and [CppReference](https://en.cppreference.com/w/). Since I am not actively developing in C++20 at the moment, I am marking this PR as a draft to review and feedback.

Changes made:
- Replaced xor_buf calls with std::span-based equivalents.
- Handled several smaller TODOs identified in the codebase.

Confirmed that builds and test suites pass with:
```
ninja clean && ./configure.py --without-documentation --cc=clang --compiler-cache=ccache --build-targets=static,cli,tests --build-tool=ninja && ninja && ./botan-test
```

Check code format with:
```
python3 src/scripts/dev_tools/run_clang_format.py --clang-format=clang-format-17 --src-dir=src
```

---
**Outstanding considerations:**
I intentionally did not modify the [following function](https://github.com/randombit/botan/blob/e1d1cb2c837f656c3b600ffa3eaa926c3f07a410/src/lib/utils/mem_ops.h#L424), as it touches multiple areas and could introduce bugs without extensive review:
```cpp
// TODO: deprecate and replace, use .subspan()
inline void xor_buf(std::span<uint8_t> out, std::span<const uint8_t> in, size_t n) {
   BOTAN_ARG_CHECK(out.size() >= n, "output span is too small");
   BOTAN_ARG_CHECK(in.size() >= n, "input span is too small");
   xor_buf(out.first(n), in.first(n));
}
```
Files affected by this function;
```
src/lib/modes/aead/eax/eax.cpp
src/lib/mac/cmac/cmac.cpp
src/lib/pubkey/dlies/dlies.cpp
```

---
**Additional notes:**
There is a change in TLS, so additional checking may be required.

Thanks in advance for your time and guidance.

KaganCanSit committed:
Update deprecated mem_ops.h func calls to use std::span-based versions and make others todo

KaganCanSit commented:
[ci / Analysis (coverage, gcc, ubuntu-24.04) (pull_request)](https://github.com/randombit/botan/actions/runs/14004841261/job/39217312363?pr=4783)  Failing after 3m`

I see that GCC handles std::span quite differently than Clang. I'll do some research on that.

reneme commented on src/lib/pubkey/pqcrystals/pqcrystals.h:
The changes in this file are independent of the rest of the pull request and need additional consideration. Please revert.

reneme commented on src/lib/tls/tls12/tls_record.cpp:
```diff
@@ -89,13 +89,13 @@ std::vector<uint8_t> Connection_Cipher_State::aead_nonce(uint64_t seq, RandomNum
          std::vector<uint8_t> nonce(12);
          store_be(seq, nonce.data() + 4);
-         xor_buf(nonce, m_nonce.data(), m_nonce.size());
+         xor_buf(std::span(nonce), std::span(m_nonce));
```
```suggestion
         xor_buf(nonce, m_nonce);
```

... vectors are ranges by itself. No need for the spans.

reneme commented on src/lib/tls/tls12/tls_record.cpp:
```diff
@@ -89,13 +89,13 @@ std::vector<uint8_t> Connection_Cipher_State::aead_nonce(uint64_t seq, RandomNum
       case Nonce_Format::AEAD_XOR_12: {
          std::vector<uint8_t> nonce(12);
          store_be(seq, nonce.data() + 4);
-         xor_buf(nonce, m_nonce.data(), m_nonce.size());
+         xor_buf(std::span(nonce), std::span(m_nonce));
          return nonce;
```
While we're on it:

```C++
std::vector<uint8_t> nonce(12);
store_be(seq, std::span{nonce}.last<8>());
xor_buf(nonce, m_nonce);
return nonce;
```

reneme commented on src/lib/tls/tls12/tls_record.cpp:
```diff
@@ -89,13 +89,13 @@ std::vector<uint8_t> Connection_Cipher_State::aead_nonce(uint64_t seq, RandomNum
       }
       case Nonce_Format::AEAD_IMPLICIT_4: {
          BOTAN_ASSERT_NOMSG(m_nonce.size() == 4);
          std::vector<uint8_t> nonce(12);
-         copy_mem(&nonce[0], m_nonce.data(), 4);
+         copy_mem(std::span<uint8_t>(nonce.data(), 4), std::span<const uint8_t>(m_nonce.data(), 4));
          store_be(seq, &nonce[nonce_bytes_from_handshake()]);
          return nonce;
```
Maybe:

```C++
BOTAN_ASSERT_NOMSG(m_nonce.size() == 4);
std::vector<uint8_t> nonce(12);
copy_mem(std::span{nonce}.first<4>(), m_nonce);
store_be(seq, std::span{nonce}.last<8>());
return nonce;
```

... my understanding is that `nonce_bytes_from_handshake()` is always equal to `m_nonce.size()` and we assert that the latter is "4" in this case. So that should be safe.

reneme replied:
Or even like that (using `stl_util.h`):

```C++
BOTAN_ASSERT_NOMSG(m_nonce.size() == 4);
return concat(m_nonce, store_be(seq));
```

KaganCanSit replied:
Obviously I think it should be safe too, but I'll take the second way. There might be something I'm not seeing.

reneme commented on src/lib/tls/tls12/tls_record.cpp:
```diff
@@ -121,7 +121,7 @@ std::vector<uint8_t> Connection_Cipher_State::aead_nonce(const uint8_t record[],
       case Nonce_Format::AEAD_XOR_12: {
          std::vector<uint8_t> nonce(12);
          store_be(seq, nonce.data() + 4);
-         xor_buf(nonce, m_nonce.data(), m_nonce.size());
+         xor_buf(std::span(nonce), std::span(m_nonce));
          return nonce;
```
```C++
std::vector<uint8_t> nonce(12);
store_be(seq, std::span{nonce}.last<8>());
xor_buf(nonce, m_nonce);
return nonce;
```

... as above.

reneme commented on src/lib/tls/tls12/tls_record.cpp:
```diff
@@ -130,8 +130,9 @@ std::vector<uint8_t> Connection_Cipher_State::aead_nonce(const uint8_t record[],
          }
          std::vector<uint8_t> nonce(12);
-         copy_mem(&nonce[0], m_nonce.data(), 4);
-         copy_mem(&nonce[nonce_bytes_from_handshake()], record, nonce_bytes_from_record());
+         copy_mem(std::span(nonce.data(), 4), std::span(m_nonce.data(), 4));
+         copy_mem(std::span(nonce.data() + nonce_bytes_from_handshake(), nonce_bytes_from_record()),
+                  std::span(record, nonce_bytes_from_record()));
          return nonce;
```
```C++
std::vector<uint8_t> nonce(12);
copy_mem(std::span{nonce}.first<4>(), m_nonce);
copy_mem(std::span{nonce}.last(nonce_bytes_from_record()),
         std::span{record, record_len}.first(nonce_bytes_from_record()));
return nonce;
```

reneme replied:
Perhaps even (using tools from `stl_util.h`):

```C++
return concat(m_nonce, std::span{record, record_len}.first(nonce_bytes_from_record()));
```

KaganCanSit replied:
Definitely cleaner and more readable. Thanks for the guidance.

reneme reviewed (commented):
Some suggestions regarding the effective usage of `std::span`.

KaganCanSit commented:
Thank you for your input and guidance. I will incorporate the suggested changes, verify the behavior through testing, and submit a review request accordingly.

KaganCanSit commented on src/lib/utils/mem_ops.h:
```diff
@@ -218,7 +218,7 @@ inline constexpr ToT typecast_copy(FromR&& src) noexcept {
 }
 
-// TODO: deprecate and replace
+BOTAN_DEPRECATED("This function is deprecated, use span version instead")
```
You are right. My wording is not correct according to the C++ standard and the structure of the functions. I have made the changes to correct it.

KaganCanSit commented on src/lib/x509/certstor_system_macos/certstor_macos.cpp:
```diff
@@ -67,10 +67,8 @@ class scoped_CFType {
-      // TODO: C++14 - use std::get<ASN1_String>(), resp. std::get<OID>()
-      const auto oid = rdn.first;
-      auto str = rdn.second;
+   for(const auto& [oid, str_val] : dn.dn_info()) {
+      auto str = str_val;  // Create a modifiable copy
```
This change really looks more readable and elegant. I didn't want to edit it to avoid making too many changes.

KaganCanSit commented on src/lib/misc/roughtime/roughtime.cpp:
```diff
@@ -204,8 +205,12 @@ Response Response::from_bits(const std::vector<uint8_t>& response, const Nonce&
    auto hash = hashLeaf(nonce.get_nonce());
    auto index = indx;
    size_t level = 0;
+
+   std::span<const uint8_t> path_span{path};
+
    while(level < levels) {
-      hashNode(hash, typecast_copy<std::array<uint8_t, 64>>(path.data() + level * 64), index & 1);
+      auto node_data = path_span.subspan(level * 64, 64);
+      hashNode(hash, typecast_copy<std::array<uint8_t, 64>>(node_data), index & 1);
       ++level;
       index >>= 1;
    }
```
The approach here is really more efficient and readable. I will also look into your question. Thank you.

reneme commented on src/lib/utils/mem_ops.h:
```diff
@@ -427,15 +427,15 @@ inline void xor_buf(std::span<uint8_t> out, std::span<const uint8_t> in, size_t
    xor_buf(out.first(n), in.first(n));
 }
```
Did you leave this function out intentionally? (I'm commenting its closing bracket due to a review limitation of GitHub.)

Could also be deprecated, in my opinion:

```C++
BOTAN_DEPRECATED("This function is deprecated, use range-based version and std::span::first")
```

KaganCanSit replied:
​In the "Outstanding Considerations" section of the Pull Request description, I indicated my intention to deprecate a specific this function. Following various adjustments, I observed that this change adversely affected multiple tests, leading to failures. Despite exploring different approaches, I remained uncertain about certain aspects. Consequently, to mitigate the risk of errors, I deemed it prudent to exclude the impacted files from this merge request.​

However, since only this function remains, I can revisit it with additional support or create a separate merge request for it after this one. This approach would allow us to review it together as a Draft Pull Request. I am more motivated to pursue a separate request for this function.

reneme replied:
Alright. Understood. Let's not touch it for now. As stated before, this entire header will go away from the public API anyway. No need to over-complicate things now. Thanks.

reneme commented on src/lib/mac/x919_mac/x919_mac.cpp:
```diff
@@ -30,12 +30,12 @@ void ANSI_X919_MAC::add_data(std::span<const uint8_t> input) {
    m_des1->encrypt(m_state);
    while(in.remaining() >= 8) {
-      xor_buf(m_state, in.take(8).data(), 8);
+      xor_buf(std::span{m_state}.first<8>(), in.take<8>());
```
Just for the record: I just realized that `m_state` is actually 8 bytes wide by definition. Technically, this should work too:

```suggestion
      xor_buf(m_state, in.take<8>());
```

... BUT this change would incur a runtime length check for every iteration, because passing a `secure_vector` does not carry the compile-time length information.

KaganCanSit replied:
You are right, if we make this change, the length check will be repeated every iteration due to typecast_copy. We made some changes accordingly in the previous review.

Such unnecessary runtime checks can lead to performance loss and conflict with the goal of optimized readable code. Therefore, we could push this check to compile time by defining a length constant constexpr and using a fixed-size span like std::span<uint8_t, N>.

However, this solution may slightly reduce the readability of the code. There is an expression in Turkish. “The stone you throw is not worth the bird you scare.” (An idiom used when the resources used to do a job are more than the benefit you will get as a result of the job). Consider the following example:

constexpr size_t STATE_SIZE = 8;
xor_buf(std::span<uint8_t, STATE_SIZE>(m_state.data(), STATE_SIZE), in.take<STATE_SIZE>());

I am not very sure about readability; I think this point should be taken into account if changes are made.

reneme replied:
Yeah, I agree. This can and should be tackled when we explicitly refactor this class. Fine by me to leave as-is (i.e. with the magic numbers) for now.

reneme commented on src/lib/misc/roughtime/roughtime.cpp:
```diff
@@ -149,7 +151,7 @@ std::array<uint8_t, N> vector_to_array(std::vector<uint8_t, T> vec) {
    if(vec.size() != N) {
       throw std::logic_error("Invalid vector size");
    }
-   return typecast_copy<std::array<uint8_t, N>>(vec.data());
+   return typecast_copy<std::array<uint8_t, N>>(vec);
```
Actually the range-based `typecast_copy` will do a length check to ensure that the input and output buffers match. So we could get rid of the error check here. Sorry, I missed that yesterday.

KaganCanSit replied:
It doesn't matter. What matters to me is the information I get from here. I'm going through it again and I'll edit it in a moment.

reneme reviewed (approved):
Looks good to me now. Thank you for taking this on!

For the record: The `mem_ops.h` header is marked as "future internal header". So, technically, we don't even need to mark the `typecast_copy` and `xor_buf` overloads as "deprecated", because for library users the entire header is going to go away in a future major release.

Personally, I still think your change in `mem_ops.h` makes sense as-is, for documentation purposes.

@randombit, over to you. :)

KaganCanSit commented:
The conflict below was resolved by re-basing the branch with the master branch.

![image](https://github.com/user-attachments/assets/bb65cf5e-7cf3-44d7-8b67-ab9b5abc2c40)

KaganCanSit commented:
Hi @reneme,

I've been reviewing this PR during my free time. While rebasing my recent PRs to the current master branch, I realized I could also address the function I had previously left out of scope. I mentioned this in the PR description and we [discussed it here](https://github.com/randombit/botan/pull/4783#discussion_r2018059308).

I worked on this after hours today and have now removed all TODO messages while implementing the deprecation notifications. The changes pass ./botan-test successfully in my local environment.
Since you've already reviewed this PR and it will require re-review, I wanted to consult with you before proceeding. Here's the [commit](https://github.com/KaganCanSit/botan/commit/4db855b7001ebb8b532bb67d96be1cc7adddb1bb) containing my changes.

I'd like to request cherry-picking these changes to this branch, which would allow us to remove all TODO messages and publish the deprecation notifications together. This PR is closer to completion than my other requests, which motivates me to finalize it.

Since I don't actively use C++20 here, I may have mistakes. Therefore, a re-check will be needed.

What are your thoughts?

Best regards.

KaganCanSit commented:
Hello again, Mr. @reneme,

I included the commit that I made improvements for the function I mentioned in the comment above, which I did not change, in the branch here. I also rebase to the master branch. I expect the tests to be successful, I will follow it.

I have not squashed the commits so that you can easily analyze the last parts I added here. Please let me know if there are any changes you want and I will try to fix them as soon as I get a chance. I think this branch is close to being merged.

After review, I will squash the commits and make them ready for merging.

Best regards.

reneme reviewed (commented):
Thanks for your patience with this. I left a few suggestions for simplification (mostly unnecessary explicit construction of `std::span`).

Regarding the changes in `xts.cpp`: I think we should postpone those until some more top-down modernization has landed. Most notably #4880 and similar effort to spanify `process_msg` and `start_msg`. Once that is done, the abstract interface of all cipher mode implementations will be purely span-based and ready for such modernizations.

KaganCanSit commented on src/lib/misc/roughtime/roughtime.cpp:
```diff
@@ -203,10 +202,10 @@ Response Response::from_bits(const std::vector<uint8_t>& response, const Nonce&
-      hashNode(hash, typecast_copy<std::array<uint8_t, 64>>(path.data() + level * 64), index & 1);
-      ++level;
+
+   BufferSlicer slicer(path);
+   for(size_t level = 0; level < levels; ++level) {
+      hashNode(hash, typecast_copy<std::array<uint8_t, 64>>(slicer.take<64>()), index & 1);
       index >>= 1;
    }
```
I agree that switching to std::span will eliminate unnecessary data duplication and improve code clarity.

As requested, I will not address this issue in the current PR. the use of typecast_copy was already present in the original implementation.

If you like, I can create a separate issue to track this modernization effort and work on it in a future PR when time permits. Alternatively, feel free to create an issue and assign it to me - I'd be happy to take a look at it.

Translated with DeepL.com (free version)

KaganCanSit replied:
The improvement mentioned here was addressed as a separate PR in #4994 and included in the project.

KaganCanSit commented:
I have addressed all feedback from the review:

- Changes to `src/lib/modes/xts/xts.cpp` have been rolled back as suggested to avoid conflicts with upcoming modernizations (#4880, #4151)
- Retained deprecation warnings in headers to prevent new ptr-length style additions.
- The existing branch has been re-baselined to the parent branch.
- Implemented other requested fixes and improvements. (Open typing is a bit my style, sorry.)

roughtime std::span modernization proposal `BufferSlicer` is noted for future work - I may address this in a separate PR once it is merged. [https://github.com/randombit/botan/pull/4783#discussion_r2202686176](Comment)

I will follow the results of the CI tests and then we can merge if you see fit. I see you're quite busy with clang-tidy today, I'll rebase it from time to time if needed, just FYI.

reneme reviewed (approved):
Looks good to me. Thanks for your efforts.

randombit reviewed (commented):
Sorry for the wait here. Changes are generally fine but the issue is mem_ops.h is currently a public header *but* I don't think the library should be in the business of providing general utility functions like this, which is why the header is deprecated and will be removed in Botan4. Any new functionality along these lines should go into the (internal) mem_utils.h header. Ideally we can update all of the callers within the library and then guard the declarations here, which are no longer used internally but which we must continue to provide for now for compatability, with `!defined(BOTAN_IS_BEING_BUILT)` as is done with various other functions, so it's not possible to add new code to the library which uses them.

KaganCanSit commented:
Hello @randombit, @reneme,

First, I wish you a healthy, peaceful, and successful New Year. I hope to contribute more to the project in 2026.

I wanted to share a note regarding this long-standing development branch. Some of my branches have been open for a long time and for some I’ll even need to revisit the details to fully recall what I did (there are ones approaching a year). For this reason, we could follow the approach @reneme [outlined in this comment](https://github.com/randombit/botan/pull/4783#issuecomment-3346427126).

Another idea I have is: if avoiding changes to the header file (`mem_ops.h`) is important for you, then there aren’t actually any new functions added here. This PR only introduces deprecation annotations. If preferred, we could update the usages within the project to use std::span without making any changes to the header. That way, future modifications will be easier and will be future for Botan4 only change header file.

I would really like to contribute to this development. Unfortunately, I don’t have much practical experience with std::span yet, as I haven’t been using the latest standard extensively. However, I am always open to your guidance and feedback.

Kind regards.

Updates:
* 25/04/26: Rebased to main branch. Resolved `test_roughtime.cpp` merge conflict and updated `tls_record.cpp` to include `concat_util.h`, as `concat()` was moved out of `stl_util.h`.
* 24/05/26: Reabase to main branch.
